sale_discount_code/models/discount_code_rule.py

```diff
@@ -0,0 +1,103 @@
+# -*- coding: utf-8 -*-
+# Copyright 2017 Akretion (http://www.akretion.com).
+# @author Benoît GUILLOT <benoit.guillot@akretion.com>
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl).
+
+from openerp import api, fields, models, _
+import openerp.addons.decimal_precision as dp
+from openerp.exceptions import Warning as UserError
+
+
+class DiscountCodeRule(models.Model):
+    _name = 'discount.code.rule'
+    _description = 'Discount Code Rule'
+
+    sequence = fields.Integer('Sequence', default=10)
+    application_type = fields.Selection(
+        selection=[('manual', 'Manual')], string='Application type',
+        default='manual')
+    code = fields.Char('Code')
+    discount_amount = fields.Float(
+        string='Discount amount',
+        digits_compute=dp.get_precision('Account'),
+        required=True)
+    date_from = fields.Date()
+    date_to = fields.Date()
+    restriction_method = fields.Selection(
+        selection=[
+            ('no_restriction', 'No restriction'),
+            ('partner_list', 'Partner list'),
+            ('newsletter', 'Only newsletter'),
+            ('pricelist', 'Pricelist')],
+        string="Restriction method")
+    restrict_partner_ids = fields.Many2many(
+        comodel_name='res.partner',
+        relation='discount_rule_partner_rel',
+        column1='rule_id',
+        column2='partner_id',
+        string='Restricted partners')
+    restrict_pricelist_ids = fields.Many2many(
+        comodel_name='product.pricelist',
+        relation='discount_rule_pricelist_rel',
+        column1='rule_id',
+        column2='pricelist_id',
+        string='Restricted pricelists')
+    usage_restriction = fields.Selection(
+        selection=[
+            ('one_per_partner', 'One per partner'),
+            ('no_restriction', 'No restriction')])
+    restriction_amount = fields.Selection(
+        selection=[
+            ('taxed_amount', 'Taxed amount'),
+            ('untaxed_amount', 'Untaxed amount')])
+    minimal_amount = fields.Float(
+        string='Minimal amount',
+        digits_compute=dp.get_precision('Account'))
+
+    _sql_constraints = [
+        ('code_unique', 'UNIQUE (code)', _('Discount code must be unique !'))]
+
+    @api.model
+    def _check_apply(self, order):
+        if (self.date_to and fields.Date.today() > self.date_to) \
+                or (self.date_from and fields.Date.today() < self.date_from):
+            return False
+        if self.minimal_amount and (
+                (self.restriction_amount == 'taxed_amount' and self.minimal_amount > order.amount_total) \
+                or (self.restriction_amount == 'untaxed_amount' and self.minimal_amount > order.amount_untaxed)):
+            return False
+        if self.restriction_method == 'partner_list' and order.partner_id.id not in rule.restrict_partner_ids.ids:
```

Comment:
maybe it will be more readable to have something like
`
def _check_retriction_partner_list(self):
    return self.partner_id.id not in self.restrict_partner_ids.ids


restrict_function = 'check_restriction_%s' % self.restriction_method
if restrict_function():
    return False
`